commit 54b87d1082e6971aebee3184d7d408ab248dc6d0
Author: Luite Stegeman <stegeman@gmail.com>
Date:   Sat Apr 4 08:48:32 2026 +0200

    rts: add IS_BLACKHOLE_INFO, IS_BLACKHOLE_OR_WHITEHOLE_INFO macros

diff --git a/rts/Messages.c b/rts/Messages.c
index 7238fa1c91..d9849ca8ee 100644
--- a/rts/Messages.c
+++ b/rts/Messages.c
@@ -176,10 +176,7 @@ uint32_t messageBlackHole(Capability *cap, MessageBlackHole *msg)
     // BLACKHOLE has already been updated, and GC has shorted out the
     // indirection, so the pointer no longer points to a BLACKHOLE at
     // all.
-    if (bh_info != &stg_BLACKHOLE_info &&
-        bh_info != &stg_CAF_BLACKHOLE_info &&
-        bh_info != &__stg_EAGER_BLACKHOLE_info &&
-        bh_info != &stg_WHITEHOLE_info) {
+    if (!IS_BLACKHOLE_OR_WHITEHOLE_INFO(bh_info)) {
         return 0;
     }
 
@@ -338,10 +335,7 @@ StgTSO * blackHoleOwner (StgClosure *bh)
 
     info = RELAXED_LOAD(&bh->header.info);
 
-    if (info != &stg_BLACKHOLE_info &&
-        info != &stg_CAF_BLACKHOLE_info &&
-        info != &__stg_EAGER_BLACKHOLE_info &&
-        info != &stg_WHITEHOLE_info) {
+    if (!IS_BLACKHOLE_OR_WHITEHOLE_INFO(info)) {
         return NULL;
     }
 
diff --git a/rts/ThreadPaused.c b/rts/ThreadPaused.c
index 484bb67e09..1410ee7a3d 100644
--- a/rts/ThreadPaused.c
+++ b/rts/ThreadPaused.c
@@ -304,9 +304,7 @@ threadPaused(Capability *cap, StgTSO *tso)
             // Note that great care is required when entering computations
             // suspended by this mechanism. See Note [AP_STACKs must be eagerly
             // blackholed] for details.
-            if (((bh_info == &stg_BLACKHOLE_info
-                  || bh_info == &__stg_EAGER_BLACKHOLE_info
-                  || bh_info == &stg_CAF_BLACKHOLE_info)
+            if ((IS_BLACKHOLE_INFO(bh_info)
                 && !threadPausedBlackHoleOwner(tso, bh))
                 || (bh_info == &stg_WHITEHOLE_info))
             {
@@ -338,16 +336,12 @@ threadPaused(Capability *cap, StgTSO *tso)
             // shouldn't overwrite its payload: There may already be a blocking
             // queue (see #26324).
             if(frame_info == &stg_bh_upd_frame_info
-               || bh_info == &stg_BLACKHOLE_info
-               || bh_info == &__stg_EAGER_BLACKHOLE_info
-               || bh_info == &stg_CAF_BLACKHOLE_info) {
+               || IS_BLACKHOLE_INFO(bh_info)) {
                 // already a black hole: we do nothing
 
                 // it should be a black hole (but we may not own it, as another
                 // thread could have raced us to claim it)
-                ASSERT(bh_info == &stg_BLACKHOLE_info ||
-                       bh_info == &__stg_EAGER_BLACKHOLE_info ||
-                       bh_info == &stg_CAF_BLACKHOLE_info);
+                ASSERT(IS_BLACKHOLE_INFO(bh_info));
 
             } else {
                 // lazy black hole
diff --git a/rts/Threads.c b/rts/Threads.c
index 706d7f63a9..61d5437476 100644
--- a/rts/Threads.c
+++ b/rts/Threads.c
@@ -442,9 +442,7 @@ checkBlockingQueues (Capability *cap, StgTSO *tso)
         // thing the result would be the same in almost all cases. See #20093.
         p = UNTAG_CLOSURE(bq->bh);
         const StgInfoTable *pinfo = ACQUIRE_LOAD(&p->header.info);
-        if ((pinfo != &stg_BLACKHOLE_info &&
-             pinfo != &__stg_EAGER_BLACKHOLE_info &&
-             pinfo != &stg_CAF_BLACKHOLE_info) ||
+        if (!IS_BLACKHOLE_INFO(pinfo) ||
             (RELAXED_LOAD(&((StgInd *)p)->indirectee) != (StgClosure*)bq))
         {
             wakeBlockingQueue(cap,bq);
@@ -468,10 +466,7 @@ updateThunk (Capability *cap, StgTSO *tso, StgClosure *thunk, StgClosure *val)
     const StgInfoTable *i;
 
     i = ACQUIRE_LOAD(&thunk->header.info);
-    if (i != &stg_BLACKHOLE_info &&
-        i != &stg_CAF_BLACKHOLE_info &&
-        i != &__stg_EAGER_BLACKHOLE_info &&
-        i != &stg_WHITEHOLE_info) {
+    if (!IS_BLACKHOLE_OR_WHITEHOLE_INFO(i)) {
         updateWithIndirection(cap, thunk, val);
         return;
     }
diff --git a/rts/include/rts/storage/ClosureMacros.h b/rts/include/rts/storage/ClosureMacros.h
index 1dcba95dd9..78c11cc0e9 100644
--- a/rts/include/rts/storage/ClosureMacros.h
+++ b/rts/include/rts/storage/ClosureMacros.h
@@ -372,6 +372,14 @@ EXTERN_INLINE StgOffset BLACKHOLE_sizeW ( void );
 EXTERN_INLINE StgOffset BLACKHOLE_sizeW ( void )
 { return sizeofW(StgInd); } // a BLACKHOLE is a kind of indirection
 
+#define IS_BLACKHOLE_INFO(info) \
+    ((info) == &stg_BLACKHOLE_info || \
+     (info) == &__stg_EAGER_BLACKHOLE_info || \
+     (info) == &stg_CAF_BLACKHOLE_info)
+
+#define IS_BLACKHOLE_OR_WHITEHOLE_INFO(info) \
+    (IS_BLACKHOLE_INFO(info) || (info) == &stg_WHITEHOLE_info)
+
 /* --------------------------------------------------------------------------
    Sizes of closures
    ------------------------------------------------------------------------*/
